Products/CMFPlone/browser/search.py

```diff
@@ -183,10 +184,27 @@ def filter_types(self, types):
         plone_utils = getToolByName(self.context, "plone_utils")
         if not isinstance(types, list):
             types = [types]
-        return plone_utils.getUserFriendlyTypes(types)
+
+        # Get the vocabulary
+        vocab_factory = self._portal_types_vocabulary()
+        if vocab_factory is None:
+            # Fallback to original behavior if vocabulary is not available
+            return plone_utils.getUserFriendlyTypes(types)
+
+        vocab = vocab_factory(self.context)
+
+        # Filter types based on plone_utils and maintain vocabulary order
+        user_friendly_types = set(plone_utils.getUserFriendlyTypes(types))
+        sorted_types = [term.value for term in vocab if term.value in user_friendly_types]
+
+        return sorted_types
+    
+    def _portal_types_vocabulary(self):
+        return queryUtility(
+            IVocabularyFactory, "plone.app.vocabularies.ReallyUserFriendlyTypes"
+        )
 
     def types_list(self):
-        # only show those types that have any content
```

Comment:
Why did you remove this comment?

Reply:
+1 please keep it.

Reply:
ah, that might be a mistake... apologies